pkg/core/region.go

```diff
@@ -1833,6 +1833,42 @@ func (r *RegionsInfo) ScanRegions(startKey, endKey []byte, limit int) []*RegionI
 	return res
 }
 
+// BatchScanRegions scans regions in given key pairs, returns at most `limit` regions.
+// limit <= 0 means no limit.
+// The given key pairs should be non-overlapping.
+func (r *RegionsInfo) BatchScanRegions(keyRanges *KeyRanges, limit int) []*RegionInfo {
+	r.t.RLock()
+	defer r.t.RUnlock()
```

Comment:
Will it hold the lock for a long time?

Reply:
Searching in memory is fast anyway, and there is a limit parameter which should be set properly by client([default 128](https://github.com/tikv/client-go/blob/ac8fa1d73a0c55e8526a1ae9b88670b1b64970e6/internal/locate/region_cache.go#L78) in client-go).

Reply:
Ok, if it holds the lock for a long time, It might affect the heartbeat process.